src/gui/properties/piecesbar.cpp

```diff
@@ -0,0 +1,325 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2016 Eugene Shalygin <eugene.shalygin@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "piecesbar.h"
+
+#include <QApplication>
+#include <QHelpEvent>
+#include <QPainter>
+#include <QTextStream>
+#include <QToolTip>
+
+#include "base/bittorrent/torrenthandle.h"
+#include "base/utils/misc.h"
+
+namespace
+{
+    class PieceIndexToImagePos
+    {
+    public:
+        PieceIndexToImagePos(const BitTorrent::TorrentHandle * const torrent, const QImage& image)
+            : m_piecesCount{torrent->info().piecesCount()}
+            , m_imageWidth{image.width()}
+        {}
+
+        int imagePos(int pieceIndex) const
+        {
+            return static_cast<float>(pieceIndex) / (m_piecesCount - 1) * (m_imageWidth - 1);
+        }
+
+        int pieceIndex(int imagePos) const
+        {
+            return static_cast<float>(imagePos) / (m_imageWidth - 1) * (m_piecesCount - 1);
+        }
+
+    private:
+        const int m_piecesCount;
+        const int m_imageWidth;
+    };
+}
+
+PiecesBar::PiecesBar(QWidget *parent)
+    : QWidget{parent}
+    , m_torrent{nullptr}
+    , m_image{}
+    , m_borderColor{palette().color(QPalette::Dark)}
+    , m_bgColor{Qt::white}
+    , m_pieceColor{Qt::blue}
+    , m_pieceColors{}
+    , m_hovered{false}
+    , m_highlitedRegion{}
+{
```

Comment:
Lets leverage the language and not be very verbose with no reason. If it has a default constructor don't empty initialize it. So pelase delete those 3.